Source/Core/DolphinWX/GameListCtrl.cpp

```diff
@@ -768,33 +754,81 @@ void GameListCtrl::RescanList()
   std::set_difference(search_results.cbegin(), search_results.cend(), cached_paths.cbegin(),
                       cached_paths.cend(), std::back_inserter(new_paths));
 
-  const Core::TitleDatabase title_database;
+  // Reload the TitleDatabase
+  m_title_database = {};
+
   // For now, only scan new_paths. This could cause false negatives (file actively being written),
   // but otherwise should be fine.
-  for (const auto& path : removed_paths)
+  bool cache_changed = false;
   {
-    auto it = std::find_if(
-        m_cached_files.cbegin(), m_cached_files.cend(),
-        [&path](const std::shared_ptr<GameListItem>& file) { return file->GetFileName() == path; });
-    if (it != m_cached_files.cend())
-      m_cached_files.erase(it);
+    std::unique_lock<std::mutex> lk(m_cache_mutex);
+    for (const auto& path : removed_paths)
+    {
+      auto it = std::find_if(m_cached_files.cbegin(), m_cached_files.cend(),
+                             [&path](const std::shared_ptr<GameListItem>& file) {
+                               return file->GetFileName() == path;
+                             });
+      if (it != m_cached_files.cend())
+      {
+        cache_changed = true;
+        m_cached_files.erase(it);
+      }
+    }
+    for (const auto& path : new_paths)
+    {
+      auto file = std::make_shared<GameListItem>(path, m_title_database);
+      if (file->IsValid())
+      {
+        cache_changed = true;
+        m_cached_files.push_back(std::move(file));
+      }
+    }
   }
-  for (const auto& path : new_paths)
+  // The common case is that just a file has been added/removed, so trigger a refresh ASAP with the
+  // assumption that other properties of files will not change at the same time (which will be fine
+  // and just causes a double refresh).
+  if (cache_changed)
+    QueueEvent(new wxCommandEvent(DOLPHIN_EVT_REFRESH_GAMELIST));
+
+  // If any cached files need updates, apply the updates to a copy and delete the original - this
+  // makes the UI thread's use of cached files safe. Note however, it is assumed that RefreshList
+  // will not iterate m_cached_files while the scan thread is modifying the list itself.
+  bool refresh_needed = false;
   {
-    auto file = std::make_shared<GameListItem>(path, title_database);
-    if (file->IsValid())
-      m_cached_files.push_back(std::move(file));
+    std::unique_lock<std::mutex> lk(m_cache_mutex);
+    for (auto it = m_cached_files.begin(); it != m_cached_files.end();)
+    {
+      auto file = *it;
+      bool emu_state_changed = false, banner_changed = false;
+      if (file->EmuStateChanged())
```

Comment:
It's written like this so the copy is not done if neither thing changed. Maybe I'm not sure what you mean since the formatting messed up.